### ISSUE TITLE
image dataset generation functions for image segmentation data

### ISSUE THREAD

**Issue by morrisalp** [tensorflow/tensorflow#49514]
**System information**
- TensorFlow version (you are using): 2.4.1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Image (semantic) segmentation tasks consist of input image data and target image masks which classify each pixel of the input image. Example: https://www.tensorflow.org/tutorials/images/segmentation

Right now the Keras image data preprocessing API has convenient functions `image_dataset_from_directory` and `ImageDataGenerator.flow_from_directory` for creating TF Datasets for image data, but these are only suitable for image classification tasks (each image has a single classification).

Using these APIs for image segmentation requires some hacky fiddling (I'm currently using this solution from [SO](https://stackoverflow.com/questions/58050113/imagedatagenerator-for-semantic-segmentation)) e.g. creating two ImageDataGenerators with the same seed and `class_mode=None`, and some directory manipulation.

I would suggest adding something like `class_mode='segmentation'` along with a standard directory format for such tasks.

**Will this change the current api? How?**

It would add `class_mode='segmentation'` to these functions. For ImageDataGenerator, it requires defining whether augmentation is applied to the input image, target mask, and/or both.

**Who will benefit with this feature?**
Everyone working on image segmentation tasks

**Any other info**
This could also affect `flow_from_dataframe()`.

**Comment by fchollet** [comment]
Thanks for opening this issue. I think we should offer a better API for image segmentation use cases (and also related use cases like object detection, keypoint detection, etc).

There are 2 subproblems:

1. Generating a dataset
2. Image data augmentation (same augmentation needs to be applied to inputs and to segmentation masks or target bounding boxes)

For 1. I don't think adding new built-in options in `image_dataset_from_directory` is the best option, because the range of possible formats for segmentation masks is large.

I'd recommend using the utility in a modular way, like so:

```python
img_ds = image_dataset_from_directory(flat_img_dir, labels=None, shuffle=None)
segmentation_mask_ds = image_dataset_from_directory(flat_mask_dir, labels=None, shuffle=None, color_mode='grayscale')  # Assumes the filenames are in the same order
ds = Dataset.zip((img_ds, segmentation_mask_ds))
```

Does that make sense? You can also add shuffling as long as you provide the same seed in both cases.

For data augmentation, we're working on something based on image preprocessing layers.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
